### PR TITLE
Adding v3/assets Endpoint

### PR DESCRIPTION
**Pull request by ngaglani-tenable**
- This PR contains the proposed base structure for code for v3 resources
- Current structure is something like
```
io/
   __init__.py
	v3/
            __init__.py
            base/
                endpoints/
	                uw.py
                iterators/
                schema/
	                analysis.py
	                exports.py
            assets.py
```

- Previous usage
```tio.assets.list()```
- Updated usage
```tio.v3.assets.search_assets()```

**Note - added the support for search asset. Test cases are not written since many actions depends on other API support.

please review the PR

### CODE REVIEW

File: tenable/io/v3/base/schema/explore/utils.py [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+                data_validate = dict(property=property, order=order)
+                data_validate = schema.load(data_validate)
+                if is_with_prop:
+                    sort_data.append(schema.dump(data_validate))
+                else:
+                    sort_data.append({
+                        property: order
+                    })
+            else:
+                raise ValidationError('Sort', '[("field_name_1", "asc"), ("field_name_2", "desc")]')
```
Comment (tomeara) [inline tenable/io/v3/base/schema/explore/utils.py:28]: It seems that you've guarded a good bit in the condition above, but I'm not completely certain that it'll cover everything. Let's say someone enters:
```
[("field_name_1", 0), ("field_name_2", 1)]
```
Are we handling that error case some other way?

**Review by tomeara — APPROVED** [review]
Looks good, just added a comment on the Validation. I'll approve though.

File: tenable/io/v3/base/schema/explore/utils.py (hunk 1 above)
Comment (sahils-tenable) [reply tenable/io/v3/base/schema/explore/utils.py:28]: @tomeara yes we made sure that only these two options will get passed from the schema else it will be blocked.
image

Comment (tomeara) [reply tenable/io/v3/base/schema/explore/utils.py:28]: Ah, sorry I missed that. Sounds great!

File: tenable/errors.py [hunk 2]
```diff
@@ -100,3 +100,9 @@ class PasswordComplexityError(APIError):
             infrastructure.  In the case of Non-Tenable.io products, is simply
             an empty string.
     '''
+
+
+class ValidationError(Exception):
```
[the remark this answers is not in the document]
Comment (sahils-tenable) [reply tenable/errors.py]: yes updated.

File: tenable/io/v3/assets.py [hunk 3]
```diff
@@ -0,0 +1,339 @@
...
+
+        Args:
+            asset_uuid (str): The unique identifier for the asset.
+
+        Returns:
+            None:
+
+        Examples:
+            >>> asset_id = '00000000-0000-0000-0000-000000000000'
+            >>> tio.v3.workbenches.asset_delete(asset_id)
```
[the remark this answers is not in the document]
Comment (ngaglani-tenable) [reply tenable/io/v3/assets.py]: Done.

**Review by SteveMcGrath — CHANGES_REQUESTED** [review]

File: tenable/io/v3/assets.py [hunk 4]
```diff
@@ -0,0 +1,354 @@
...
+        '''
+        return self._delete(uuid)
+
+    def details(self, uuid: UUID) -> Dict:  # pylint: disable=arguments-renamed
```
[the remark this answers is not in the document]
Comment (sahils-tenable) [reply tenable/io/v3/assets.py]: fixed

File: tenable/io/v3/assets.py [hunk 5]
```diff
@@ -0,0 +1,354 @@
...
+
+        Requires -
+            fields -- list of string = ['field1', 'field2'] -> fields is not supported by the search_assets api
+            filter -- tuple ('field_name', 'operator', 'value') -- ('and', ('test', 'oper', '1'), ('test', 'oper', '2'))
+            sort -- 'sort': [
+                        {'last_observed': 'desc'}
+                    ]
+            limit -- integer = (10)
+            next -- str = ('adfj3u4j34u9j48wi3j5w84jt5') -> next token
+        '''
```
[the remark this answers is not in the document]
Comment (sahils-tenable) [reply tenable/io/v3/assets.py]: updated

**Comment by tomeara** [comment]
It is looking pretty good, but you'll need to write up some tests; especially around the CSV iterator.

File: tenable/errors.py [hunk 6]
```diff
@@ -36,12 +37,13 @@ def __init__(self, resource: str, resource_id: str, filename: str):
         self.resource = str(resource)
         self.resource_id = str(resource_id)
         self.filename = str(filename)
-        self.msg = (f'resource {resource}:{resource_id} '
-                    f'requested file {filename} and has failed.'
-                    )
+        self.msg = (
+            f'resource {resource}:{resource_id} '
+            f'requested file {filename} and has failed.'
+        )
```
Comment (SteveMcGrath) [inline tenable/errors.py:43]: This change is unrelated to the assets API

File: tenable/io/v3/base/schema/explore/search.py [hunk 7]
```diff
@@ -13,15 +13,17 @@ class SortSchema(Schema):
...
-    filter = marshm_fields.Nested(FilterSchema)
-    limit = marshm_fields.Int()
-    next = marshm_fields.Int()
-    sort = marshm_fields.List(marshm_fields.Nested(SortSchema))
+    fields = marshm_fields.List(marshm_fields.Str(), allow_none=True,
+                                default=None)
+    filter = marshm_fields.Nested(FilterSchema, allow_none=True, default=None)
+    limit = marshm_fields.Int(default=1000)
+    next = marshm_fields.Str(allow_none=True)
+    sort = marshm_fields.List(marshm_fields.Dict(), allow_none=True,
```
Comment (SteveMcGrath) [inline tenable/io/v3/base/schema/explore/search.py:28]: why are marshmallow fields being imported as "marshmallow_fields"?  This is inconsistent with other code already in pyTenable mainline.

Comment (SteveMcGrath) [inline tenable/io/v3/base/schema/explore/search.py:28]: Shouldn't this be 
```python
sort = fields.List(fields.Nested(SortSchema))
```

File: tenable/io/v3/base/schema/explore/utils.py [hunk 8]
```diff
@@ -0,0 +1,35 @@
...
+                    sort_data.append({property: order})
+            else:
+                raise RequestValidationError(
+                    'Sort', '[("field_name_1", "asc"),'
+                            ' ("field_name_2", "desc")]'
+                )
+    else:
+        sort_data = None
+
+    return sort_data
```
Comment (SteveMcGrath) [inline tenable/io/v3/base/schema/explore/utils.py:35]: This should be a pre_load method as part of the SortSchema instead.

```python
class SortSchema(Schema):
    @pre_load
    def tuple_conversion(self, data, **kwargs):
        if is instance(data, tuple):
            data = {'property': data[0], 'order': data[1]}
        return data
```

Refer to https://github.com/dhvanilshah-tenable/pyTenable/blob/8991bcdf5703ee4a514468ef1aa1119df38693ae/tenable/io/v3/base/schema/explore/filters.py as a good example of this.

File: tenable/errors.py [hunk 9]
```diff
@@ -100,3 +102,12 @@ class PasswordComplexityError(APIError):
...
     '''
+
+
+class RequestValidationError(Exception):
+    def __init__(self, type, example):
+        message = (
+            f'{type}: Validation failed. Please provide'
+            f' the valid data. Example: {example}'
+        )
+        super().__init__(message)
```
Comment (SteveMcGrath) [inline tenable/errors.py:113]: Unnecessary is the validation was handled within marshmallow.

File: tenable/io/v3/base/endpoints/explore.py [hunk 10]
```diff
@@ -32,36 +36,72 @@ def details(self, obj_id: Union[str, UUID]):
...
-            iterator_cls=None,
-            schema_cls: Optional[Type[SearchSchema]] = None,
+            resource: str,
+            api_path: str,
+            is_sort_with_prop: bool = True,
+            return_resp: bool = False,
+            iterator_cls: Optional[SearchIterator] = SearchIterator,
+            schema_cls: Optional[SearchSchema] = SearchSchema,
             **kwargs
     ):
```
Comment (SteveMcGrath) [inline tenable/io/v3/base/endpoints/explore.py:46]: No return datatype specified

File: tenable/io/v3/base/endpoints/explore.py [hunk 11]
```diff
@@ -77,29 +117,24 @@ def search(
...
                 object is instead returned instead of an iterable.
 
-        Examples:
-
-            >>>
         '''
-        kwargs['fields'] = fields
+        sort = kwargs.get('sort')
+        next_token = kwargs.get('next')
+        sort = generate_sort_data(sort, is_sort_with_prop)
```
Comment (SteveMcGrath) [inline tenable/io/v3/base/endpoints/explore.py:123]: should be done within the schema

File: tenable/io/v3/base/schema/explore/search.py [hunk 12]
```diff
@@ -13,15 +13,17 @@ class SortSchema(Schema):
...
     '''
     Schema supporting the search request
     '''
-    fields = marshm_fields.List(marshm_fields.Str())
-    filter = marshm_fields.Nested(FilterSchema)
-    limit = marshm_fields.Int()
-    next = marshm_fields.Int()
-    sort = marshm_fields.List(marshm_fields.Nested(SortSchema))
+    fields = marshm_fields.List(marshm_fields.Str(), allow_none=True,
+                                default=None)
```
Comment (SteveMcGrath) [inline tenable/io/v3/base/schema/explore/search.py:24]: default should be dump_default and should only be specified if a None result is to be returned (instead of omission)

https://marshmallow.readthedocs.io/en/stable/marshmallow.fields.html#marshmallow.fields.Field

File: tenable/io/v3/base/schema/explore/search.py [hunk 13]
```diff
@@ -13,15 +13,17 @@ class SortSchema(Schema):
...
+    filter = marshm_fields.Nested(FilterSchema, allow_none=True, default=None)
```
Comment (SteveMcGrath) [inline tenable/io/v3/base/schema/explore/search.py:25]: default should be dump_default and should only be specified if a None result is to be returned (instead of omission)

https://marshmallow.readthedocs.io/en/stable/marshmallow.fields.html#marshmallow.fields.Field

File: tenable/io/v3/schema.py [hunk 14]
```diff
@@ -1,13 +1,82 @@
 '''
-Users API Endpoints Schemas
+V3 API Endpoints Schemas
 '''
-from marshmallow import Schema, fields
+from marshmallow import Schema, ValidationError, fields
+from marshmallow import validate as v
```
Comment (SteveMcGrath) [inline tenable/io/v3/schema.py:5]:
```python
from marshmallow import Schema, ValidationError, fields, validate as v
```

File: tenable/io/v3/schema.py [hunk 15]
```diff
@@ -1,13 +1,82 @@
...
+        self, data, **kwargs
+    ):
+        '''
+        Transform a Tuple of Assets into a List of Assets
+        and validate each Asset.
+        '''
+        data['assets'] = list(data.get('assets', []))
+
+        if len(data['assets']) < 1:
+            raise ValidationError('Provide at least one asset')
```
Comment (SteveMcGrath) [inline tenable/io/v3/schema.py:38]: Doesn't this replicate the Length validation?

File: tenable/io/v3/schema.py [hunk 16]
```diff
@@ -1,13 +1,82 @@
...
+                or asset.get('ipv4')
+                or asset.get('netbios_name')
+                or asset.get('mac_address')
+            ):
+                raise ValidationError(
+                    'Each asset object requires a value for at least one '
+                    'of the following properties: '
+                    'fqdn, ipv4, netbios_name, mac_address.'
+                )
+        return data
```
Comment (SteveMcGrath) [inline tenable/io/v3/schema.py:52]: Shouldn't this be handled with an appropriate asset schema and validation method?

```python
class AssetSchema(Schema):
    class Meta:
        unknown = INCLUDE
    
    @validates_schema
    def validate_required_attributes_exist(self, data, **kwargs):
        if not (data.get('fqdn') 
                or data.get('ipv4') 
                or data.get('netbios_name') 
                or data.get('mac_address')
                ):
            raise ValidationError('Each asset object requires a value for at '
                                  'least one of the following properties: '
                                  'fqdn, ipv4, netbios_name, mac_address.'
                                  )


class ImportAssetSchema(Schema):
    assets = fields.List(fields.nested(AssetSchema), 
                         validate=v.Length(min=1),
                         required=True
                         )
    source = fields.Str(required=True)
```

File: tenable/io/v3/base/schema/explore/utils.py [hunk 17]
```diff
@@ -0,0 +1,35 @@
...
+                    sort_data.append(schema.dump(data_validate))
+                else:
+                    sort_data.append({property: order})
+            else:
+                raise RequestValidationError(
+                    'Sort', '[("field_name_1", "asc"),'
+                            ' ("field_name_2", "desc")]'
+                )
+    else:
+        sort_data = None
```
Comment (SteveMcGrath) [inline tenable/io/v3/base/schema/explore/utils.py:33]: this should be part of the schema

File: tenable/io/v3/base/schema/explore/search.py [hunk 18]
```diff
@@ -13,15 +13,17 @@ class SortSchema(Schema):
...
+    limit = marshm_fields.Int(default=1000)
+    next = marshm_fields.Str(allow_none=True)
+    sort = marshm_fields.List(marshm_fields.Dict(), allow_none=True,
+                              default=None)
```
Comment (SteveMcGrath) [inline tenable/io/v3/base/schema/explore/search.py:29]: should be

```python
sort = fields.List(marshm_fields.Nested(SortSchema))
```

File: tenable/io/v3/base/schema/explore/search.py [hunk 19]
```diff
@@ -13,15 +13,17 @@ class SortSchema(Schema):
     Schema for the sorting sub-object
     '''
     property = marshm_fields.Str()
-    order = marshm_fields.Str(validates=v.OneOf(['asc', 'desc']))
+    order = marshm_fields.Str(validate=v.OneOf(['asc', 'desc']))
```
Comment (SteveMcGrath) [inline tenable/io/v3/base/schema/explore/search.py:17]: Sort schema should have a pre_load method for tuple conversion:

```python
class SortSchema(Schema):
    @pre_load
    def tuple_expansion(self, data, **kwargs):
        if isinstance(data, tuple):
            data = {'property': data[0]
                    'order': data[1]
                    }
        elif isinstance(data, tuple) and len(data) != 2:
            raise ValidationError('Sort tuple must be only two elements '
                                  '("property", "order")'
                                  )
        return data
```

File: tenable/io/v3/base/endpoints/explore.py [hunk 20]
```diff
@@ -32,36 +36,72 @@ def details(self, obj_id: Union[str, UUID]):
 
     def search(
             self,
-            *,
```
Comment (SteveMcGrath) [inline tenable/io/v3/base/endpoints/explore.py:35]: `*` is used to force keyword usage.

File: tenable/io/v3/base/endpoints/explore.py [hunk 21]
```diff
@@ -77,29 +117,24 @@ def search(
...
-        if not schema_cls:
-            schema_cls = SearchSchema
-        if not iterator_cls:
-            # todo - commenting this temporarily
-            # iterator_cls = ExploreSearchIterator
-            pass
-        schema = schema_cls()
-        payload = schema.dump(schema.load(kwargs))
+        if not next_token:
+            schema = schema_cls(unknown=EXCLUDE)
```
Comment (SteveMcGrath) [inline tenable/io/v3/base/endpoints/explore.py:126]: This usage should be discouraged as it allows for invalid attributes to be silently dropped.

File: tenable/io/v3/base/endpoints/explore.py [hunk 22]
```diff
@@ -77,29 +117,24 @@ def search(
...
+            payload = schema.dump(schema.load(kwargs))
+        else:
+            payload = {'next': next_token}
         if return_resp:
-            return self._post('search', json=payload)
+            return self._api.post(api_path, json=payload).json()
```
Comment (SteveMcGrath) [inline tenable/io/v3/base/endpoints/explore.py:131]: should use the `conv_json=True` parameter over `.json()`

**Review by SteveMcGrath — CHANGES_REQUESTED** [review]

**Comment by SteveMcGrath** [comment]
The iterators as a whole could be refactored to more appropriately comply with DRY.

```python
from copy import copy
from typing import Any, Dict, List, TextIO
from requests import Response
from tenable.io.v3.base.iterators.iterator import APIResultIterator


class ExploreIterator(APIResultIterator):
    _path: str = None
    _next_token: str = None
    _resource: str = None
    _payload: Dict = {}
    _headers: Dict = {}
    
    def _construct_headers(self) -> Dict:
        '''
        Constructs the headers for API calls
        '''
        return copy(self._headers)
    
    def _construct_payload(self) -> Dict:
        '''
        Constructs the payload for the API
        '''
        payload = copy(self._payload)
        if self._next_token:
            payload['next'] = self._next_token
        return payload
    
    def _process_response(self, response: Response) -> None:
        '''
        Processes the API response
        '''
        pass
    
    def _get_page(self) -> None:
        '''
        Request the next page of data
        '''
        payload = self._construct_payload()
        headers = self._construct_headers()
        resp = self._api.post(self._path,
                              json=payload,
                              headers=headers
                              )
        self._next_token = resp.headers.get('X-Continuation-Token')
        self._process_response(resp)


class SearchIterator(ExploreIterator):
    def _process_response(self, response: Response) -> None:
        '''
        Process the API Response
        '''
        body = response.json()
        pagination = body.get('pagination', {})
        self.page = body[self._resource]
        self.total = pagination.get('total')
        self._next_token = pagination.get('next')


class CSVChunkIterator(ExploreIterator):
    _headers = {'Accept': 'text/csv'}
    row_headers: str = None
    
    def _process_response(self, response: Response) -> None:
        '''
        Process the API Response
        '''
        if not self.row_headers:
            self.row_headers, _, data = response.text.partition('\n')
            self.page = [data]
        else:
            self.page = [response.text]
    
    def __getitem__(self, key: int) -> str:
        '''
        Returns the specified item
        '''
        return '\n'.join([self.row_headers, self.page[key]])
```

File: tenable/io/v3/base/schema/explore/search.py (hunk 7 above)
Comment (sahils-tenable) [reply tenable/io/v3/base/schema/explore/search.py:28]: if we import marshmallow fields in its original form then it will create a conflict with the available fields parameter for schema validation.

File: tenable/errors.py (hunk 6 above)
Comment (sahils-tenable) [reply tenable/errors.py:43]: this had been fixed by pre-commit hooks

File: tenable/io/v3/schema.py (hunk 16 above)
Comment (sahils-tenable) [reply tenable/io/v3/schema.py:52]: yes

File: tenable/io/v3/base/schema/explore/search.py [hunk 23]
```diff
@@ -1,29 +1,49 @@
...
+    order = fields.Str(validate=v.OneOf(['asc', 'desc']))
+
+    @pre_load(pass_many=True)
+    def transform_data(self, data, **kwargs):
+        if isinstance(data, tuple) and len(data) == 2:
+            property = data[0]
+            order = data[1]
+            return dict(property=property, order=order)
+        elif isinstance(data, dict):
+            return data
```
Comment (SteveMcGrath) [inline tenable/io/v3/base/schema/explore/search.py:24]: This should be unnecessary, as a validation error would be thrown by marshmallow specific to the instance type

File: tests/io/v3/base/test_search_schema.py [hunk 24]
```diff
@@ -4,18 +4,17 @@
...
 search_data = dict(
     fields=['bios_name', 'name'],
     filter=('bios_name', 'eq', 'SCCM'),
     limit=10,
-    sort=[dict(property='bios_name', order='asc')],
+    sort=[('name', 'asc'), ('bios_name', 'desc')],
     next='sdf000dfssdSDFSDFSFE00dfsdffaf'
 )
-sort_data = [('name', 'asc'), ('bios_name', 'desc')]
+sort_data = [('bios_name', 'desc'), ('name', 'asc')]
```
Comment (SteveMcGrath) [inline tests/io/v3/base/test_search_schema.py:17]: Might as well test both formats.

```python
sort_data = [('bios_name', 'desc'), {'property': 'name', 'order': 'asc'}]
```

**Comment by SteveMcGrath** [comment]
As a whole, there are is a lot of code that isn't tested:

```
tenable/io/v3/assets.py                               46     21    54%   119-121, 146, 166, 196-202, 222, 274-277, 293, 313, 340-348, 395, 400, 403
324
tenable/io/v3/base/__init__.py                         0      0   100%
325
tenable/io/v3/base/endpoints/explore.py               26     10    62%   35, 128, 139-146
326
tenable/io/v3/base/iterators/explore_iterator.py      42      1    98%   35
327
tenable/io/v3/base/iterators/iterator.py               3      0   100%
328
tenable/io/v3/base/schema/explore/filters.py          39      5    87%   52, 86, 116, 125, 132
329
tenable/io/v3/base/schema/explore/search.py           26      2    92%   26, 35
```

**Comment by sahils-tenable** [comment]
> is a lot of code that isn't tested:

For some of the asset endpoints, the endpoints needs to be build and hence response structure is unknown. So, all we will cover these tests in a separate PR.

File: tenable/io/v3/schema.py (hunk 14 above)
Comment (sahils-tenable) [reply tenable/io/v3/schema.py]: @SteveMcGrath this is being fixed by the pre-commits pipeline (isort).

File: tenable/io/v3/base/schema/explore/search.py [hunk 25]
```diff
@@ -1,27 +1,42 @@
...
 class SearchSchema(Schema):
     '''
     Schema supporting the search request
     '''
-    fields = marshm_fields.List(marshm_fields.Str())
-    filter = marshm_fields.Nested(FilterSchema)
-    limit = marshm_fields.Int()
-    next = marshm_fields.Int()
-    sort = marshm_fields.List(marshm_fields.Nested(SortSchema))
+    fields_ = fields.List(fields.Str(), allow_none=True, data_key='fields')
```
Comment (SteveMcGrath) [inline tenable/io/v3/base/schema/explore/search.py:38]: This shouldn't be necessary.  While you do have 2 things with the same name, they do not conflict.

fields != SearchSchema.fields

**Comment by SteveMcGrath** [comment]
> > is a lot of code that isn't tested:
> 
> For some of the asset endpoints, the endpoints needs to be build and hence response structure is unknown. So, all we will cover these tests in a separate PR. Also in explore.py some methods like search_results, details we have not yet started using, so once we start using these, we will write the test cases for them as well.

For the methods that simply raise a not-implemented error, I can understand that.  for the explore iterator however that should be tested out as it doesn't directly rely not he API to perform those tests.  you should be able to test out the iterators independently.

In general, I'd expect to see iterators and schemas almost always fully tested out in a PR.   

I'd also argue that you should likely test the not-implemented exceptions as well, if for no other reason than to act as reminder to write the tests when you do get to those.

File: tenable/io/v3/base/schema/explore/search.py (hunk 25 above)
Comment (sahils-tenable) [reply tenable/io/v3/base/schema/explore/search.py:38]: image
@SteveMcGrath if we use the fields as it is, then it throws error because fields variable is being re-assigned with list instead of marshmallow fields.

Comment (sahils-tenable) [reply tenable/io/v3/base/schema/explore/search.py:38]: image

Comment (sahils-tenable) [reply tenable/io/v3/base/schema/explore/search.py:38]: image

Comment (SteveMcGrath) [reply tenable/io/v3/base/schema/explore/search.py:38]: interesting, alright.  fields_ it is then.

File: tenable/io/v3/assets/api.py [hunk 26]
```diff
@@ -0,0 +1,403 @@
...
+        search_api_path = 'api/v3/assets/search'
+        return_csv = kw.pop('return_csv')
+        return super().search(
+            resource='assets',
+            iterator_cls=AssetCSVIterator if return_csv else
+            AssetSearchIterator,
+            is_sort_with_prop=False,
+            api_path=search_api_path,
+            **kw
+        )
```
Comment (SteveMcGrath) [inline tenable/io/v3/assets/api.py:128]: This could be re-written to be be more clear.

```suggestion
        iclass = AssetSearchIterator
        if kw.pop('return_csv', False):
            iclass = AssetCSVIterator
        return super().search(resource='assets',
                              iterator_cls=iclass,
                              is_sort_with_prop=False,
                              api_path=f'{self._path}/search',
                              **kw
                              )
```

File: tenable/io/v3/assets/api.py [hunk 27]
```diff
@@ -0,0 +1,403 @@
...
+        Examples:
+            >>> asset = tio.v3.assets.details(
+            ...     '00000000-0000-0000-0000-000000000000')
+        '''
+        return self._get(f'{uuid}')
+
+    def assign_tags(
+            self, action: Literal['add', 'remove'],
+            assets: List[UUID], tags: List[UUID]
+    ) -> Dict:
```
Comment (SteveMcGrath) [inline tenable/io/v3/assets/api.py:171]: currently is an anti-pattern for python.  the following below are common.  we typically prefer the first one in the mainline codebase.


```suggestion
    def assign_tags(self, 
                    action: Literal['add', 'remove'],
                    assets: List[UUID], 
                    tags: List[UUID]
                    ) -> Dict:
```

or

```suggestion
    def assign_tags(
        self, 
        action: Literal['add', 'remove'],
        assets: List[UUID], 
        tags: List[UUID]
    ) -> Dict:
```

File: tenable/io/v3/assets/api.py [hunk 28]
```diff
@@ -0,0 +1,403 @@
...
+        Examples:
+            >>> job = tio.v3.assets.import_job_details(
+            ...     '00000000-0000-0000-0000-000000000000')
+            >>> pprint(job)
+        '''
+        return self._get(f'import/jobs/{uuid}')
+
+    def move_assets(
+            self, source: UUID, destination: UUID, targets: List[str]
+    ) -> int:
```
Comment (SteveMcGrath) [inline tenable/io/v3/assets/api.py:317]: currently is an anti-pattern for python.

```suggestion
    def move_assets(self, 
                    source: UUID, 
                    destination: UUID, 
                    targets: List[str]
                    ) -> int:
```

File: tenable/io/v3/assets/api.py [hunk 29]
```diff
@@ -0,0 +1,403 @@
...
+                {'source': source,
+                 'destination': destination, 'targets': targets}
+            )
+        )
+
+        return self._patch(json=payload)['response']['data']['asset_count']
+
+    def bulk_delete(
+            self, *filters: Tuple[str], filter_type: str = None
+    ) -> Dict:
```
Comment (SteveMcGrath) [inline tenable/io/v3/assets/api.py:352]:
```suggestion
    def bulk_delete(self, 
                    *filters: Tuple[str], 
                    filter_type: str = None
                    ) -> Dict:
```

**Comment by SteveMcGrath** [comment]
Fails installable checks:

https://github.com/dhvanilshah-tenable/pyTenable/runs/4536012101?check_suite_focus=true#step:4:91

File: tenable/io/v3/base/endpoints/explore.py [hunk 30]
```diff
@@ -77,28 +119,17 @@ def search(
                 If ``return_json`` was set to ``True``, then a response
                 object is instead returned instead of an iterable.
 
-        Examples:
-
-            >>>
```
Comment (SteveMcGrath) [inline tenable/io/v3/base/endpoints/explore.py:82]: no example

File: tenable/io/v3/base/endpoints/explore.py [hunk 31]
```diff
@@ -77,28 +119,17 @@ def search(
...
-        kwargs['sort'] = sort
-        kwargs['filter'] = filter
-        kwargs['limit'] = limit
-        kwargs['next'] = next
-        if not schema_cls:
-            schema_cls = SearchSchema
-        if not iterator_cls:
-            # todo - commenting this temporarily
-            # iterator_cls = ExploreSearchIterator
-            pass
```
Comment (SteveMcGrath) [inline tenable/io/v3/base/endpoints/explore.py:94]: This should likely be removed as you are specifying the classes as default above.

File: tenable/io/v3/base/endpoints/explore.py (hunk 30 above)
Comment (sahils-tenable) [reply tenable/io/v3/base/endpoints/explore.py:82]: we have added the examples in the endpoints search function.
since this is not directly accessed by the user and having the similar representation on both, so we can avoid this.
 https://github.com/dhvanilshah-tenable/pyTenable/pull/12/files/5ad8d493343c0bdb08abdf5a755b7c154b5ec1d3#:~:text=of%20an%20iterable.-,Examples%3A,%27%27%27,-iclass%20%3D%20AssetSearchIterator

File: tenable/io/v3/base/endpoints/explore.py (hunk 31 above)
Comment (sahils-tenable) [reply tenable/io/v3/base/endpoints/explore.py:94]: these are removed from the code base.

File: tenable/io/v3/base/endpoints/explore.py [hunk 32]
```diff
@@ -33,35 +37,73 @@ def details(self, obj_id: Union[str, UUID]):
...
-            next: Optional[str] = None, return_resp: bool = False,
-            iterator_cls=None,
-            schema_cls: Optional[Type[SearchSchema]] = None,
+            resource: str,
+            api_path: str,
+            is_sort_with_prop: bool = True,
+            return_resp: bool = False,
+            iterator_cls: Optional[Union[SearchIterator, CSVChunkIterator]] =
+            SearchIterator,
+            schema_cls: Optional[SearchSchema] = SearchSchema,
```
Comment (SteveMcGrath) [inline tenable/io/v3/base/endpoints/explore.py]: 1. Optionals should only be used to convey the type when the default option would violate the typing (NoneType, for example)
2. Instead of a more complex Union, you should simply set the Typing to look for the parent class.  as both the CSV and Search iterators are children of the ExportIterator, that should be the type class.

```suggestion
            iterator_cls: ExploreIterator = SearchIterator,
            schema_cls: SearchSchema = SearchSchema,
```
